hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/ChecksumFileSystem.java

```diff
@@ -261,15 +270,129 @@ protected int readChunk(long pos, byte[] buf, int offset, int len,
           len = Math.min(len, bytesPerSum * (sumLenRead / CHECKSUM_SIZE));
         }
       }
-      if(pos != datas.getPos()) {
+      if (pos != datas.getPos()) {
         datas.seek(pos);
       }
       int nread = readFully(datas, buf, offset, len);
       if (eof && nread > 0) {
-        throw new ChecksumException("Checksum error: "+file+" at "+pos, pos);
+        throw new ChecksumException("Checksum error: " + file + " at " + pos, pos);
       }
       return nread;
     }
+
+    public static long findChecksumOffset(long dataOffset,
+                                          int bytesPerSum) {
+      return HEADER_LENGTH + (dataOffset/bytesPerSum) * FSInputChecker.CHECKSUM_SIZE;
+    }
+
+    /**
+     * Find the checksum ranges that correspond to the given data ranges.
```

Comment:
Why what is needed? You mean the code to compare the checksums? The current code requires a lot of context that isn't true in the new API. The current code is super inefficient because it did a bad job of working around those limitations. In particular, if you look at the current pread code, it reopens the crc file for each seek.